### PR TITLE
Resource folders

### LINKED ISSUE #365: Folders in resource view
We want to be able to create folders in the resource view, so not all e.g. models are just lined up. Maybe they are divided into Mushroom and Stone models. So we want to be able to divide them.

### PR DESCRIPTION
**Pull request by Chainsawkitten**
Resources can be put in folders, which are created by the user.

Fixes #365

### CODE REVIEW

File: src/Editor/GUI/FolderNameWindow.hpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <functional>
+#include <string>
+
+namespace GUI {
+    /// A window that asks the user for the name of the folder to create.
+    class FolderNameWindow {
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/FolderNameWindow.hpp]: MINOR The class 'FolderNameWindow' does not have a constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AnoConstructor)

File: src/Editor/Resources.hpp [hunk 2]
```diff
@@ -65,9 +105,16 @@ class ResourceList {
          */
         std::string GetSavePath() const;
         
+        /// The id of the next script to create.
+        unsigned int scriptNumber = 0U;
+        
     private:
         static ResourceList& GetInstance();
```
Comment (NinjaDanz3r) [inline src/Editor/Resources.hpp]: MINOR Unused private function: 'ResourceList::GetInstance' [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedPrivateFunction)

File: src/Editor/GUI/ResourceSelector.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "../Resources.hpp"
+
+namespace GUI {
+    /// Used to select a resource from the resource list.
+    class ResourceSelector {
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/ResourceSelector.hpp]: MINOR The class 'ResourceSelector' does not have a constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AnoConstructor)

File: src/Editor/Resources.cpp [hunk 4]
```diff
@@ -1,18 +1,36 @@
...
+    case Type::SCRIPT:
+        return script->name;
+    default:
+        return "";
+    }
+}
+
 ResourceList::ResourceList() {
-    activeScene = -1;
+    activeScene = "";
```
Comment (NinjaDanz3r) [inline src/Editor/Resources.cpp]: MINOR Variable 'activeScene' is assigned in constructor body. Consider performing initialization in initialization list. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuseInitializationList)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 8 issues
* ![MINOR][MINOR] 8 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [SceneEditor.cpp#L20](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/9291a3f1a4e7a965c1fa2b275175de785392c1ed/src/Editor/GUI/Editors/SceneEditor.cpp#L20): Member variable 'SceneEditor::sceneSelector' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/9291a3f1a4e7a965c1fa2b275175de785392c1ed/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L529](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/9291a3f1a4e7a965c1fa2b275175de785392c1ed/src/Engine/Manager/ScriptManager.cpp#L529): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 8 issues
* ![MINOR][MINOR] 8 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [SceneEditor.cpp#L20](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/c55cc739231013eae7df0700e3c49c7c9c226b57/src/Editor/GUI/Editors/SceneEditor.cpp#L20): Member variable 'SceneEditor::sceneSelector' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/c55cc739231013eae7df0700e3c49c7c9c226b57/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L559](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/c55cc739231013eae7df0700e3c49c7c9c226b57/src/Engine/Manager/ScriptManager.cpp#L559): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Entity/Entity.cpp [hunk 5]
```diff
@@ -100,36 +99,31 @@ Entity* Entity::InstantiateScene(const std::string& name, const std::string& ori
...
     Json::Value children = root["children"];
 
     // Loops through all the scene names.
     for (unsigned int i = 0; i < root["children"].size(); ++i) {
         if (root["children"][i]["scene"].asBool()) {
             printf("%s", root["children"][i]["sceneName"].asString().c_str());
 
-            if (originScene == root["children"][i]["sceneName"].asString())
-            {
+            if (originScene == root["children"][i]["sceneName"].asString()) {
```
Comment (SwedishSock) [inline src/Engine/Entity/Entity.cpp]: No curly brackets if if statement is only one row long.

File: src/Engine/Entity/Entity.cpp [hunk 6]
```diff
@@ -100,36 +99,31 @@ Entity* Entity::InstantiateScene(const std::string& name, const std::string& ori
...
                 error = true;
             }
-            if (error)
-            {
+            if (error) {
```
Comment (SwedishSock) [inline src/Engine/Entity/Entity.cpp]: No curly brackets if if statement is only one row long.

File: src/Engine/Entity/Entity.cpp [hunk 7]
```diff
@@ -100,36 +99,31 @@ Entity* Entity::InstantiateScene(const std::string& name, const std::string& ori
...
                 break;
             }
         }
 
-        if (!error)
-        {
+        if (!error) {
```
Comment (SwedishSock) [inline src/Engine/Entity/Entity.cpp]: No curly brackets if if statement is only one row long.

File: src/Editor/GUI/Editors/SceneEditor.cpp [hunk 8]
```diff
@@ -70,12 +68,13 @@ void SceneEditor::Show() {
...
-void SceneEditor::SetScene(std::size_t sceneIndex) {
+void SceneEditor::SetScene(const std::string& path, std::string* scene) {
     entityEditor.SetVisible(false);
-    Resources().activeScene = sceneIndex;
+    this->path = path;
+    this->scene = scene;
     
-    if (Resources().activeScene < Resources().scenes.size()) {
-        strcpy(name, Resources().scenes[Resources().activeScene].c_str());
+    if (scene != nullptr) {
```
Comment (SwedishSock) [inline src/Editor/GUI/Editors/SceneEditor.cpp]: No curly brackets if if statement is only one row long.

File: src/Editor/Resources.cpp [hunk 9]
```diff
@@ -74,63 +68,142 @@ void ResourceList::Load() {
...
+    Json::Value node;
+    node["name"] = folder.name;
     
-    // Load sounds.
-    const Json::Value soundsNode = root["sounds"];
-    for (unsigned int i = 0; i < soundsNode.size(); ++i) {
-        sounds.push_back(Managers().resourceManager->CreateSound(soundsNode[i].asString()));
+    // Save subfolders.
+    Json::Value subfolders;
+    for (const ResourceFolder& subfolder : folder.subfolders) {
```
Comment (SwedishSock) [inline src/Editor/Resources.cpp]: No curly brackets if for statement is only one row long.

File: src/Editor/Resources.cpp [hunk 10]
```diff
@@ -74,63 +68,142 @@ void ResourceList::Load() {
...
-    modelNumber = 0U;
     
-    for (TextureAsset* texture : textures) {
-        Managers().resourceManager->FreeTextureAsset(texture);
+    return folder;
+}
+
+void ResourceList::ClearFolder(ResourceFolder& folder) {
+    // Clear subfolders.
+    for (ResourceFolder& subfolder : folder.subfolders) {
```
Comment (SwedishSock) [inline src/Editor/Resources.cpp]: No curly brackets if if statement is only one row long.

File: src/Editor/Resources.hpp [hunk 11]
```diff
@@ -20,6 +21,51 @@ class ResourceList {
...
+            Geometry::Model* model;
+            
+            /// Texture.
+            TextureAsset* texture;
+            
+            /// Sound.
+            Audio::SoundBuffer* sound;
+            
+            /// Script.
+            ScriptFile* script;
```
Comment (NinjaDanz3r) [inline src/Editor/Resources.hpp]: MINOR Unused private function: 'ResourceList::GetInstance' [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedPrivateFunction)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8859be1152c0026ebf05c0373dabd2bde4f1d0c8/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L529](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8859be1152c0026ebf05c0373dabd2bde4f1d0c8/src/Engine/Manager/ScriptManager.cpp#L529): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/GUI/FolderNameWindow.hpp (hunk 1 above)
Comment (NinjaDanz3r) [inline src/Editor/GUI/FolderNameWindow.hpp]: MINOR The class 'FolderNameWindow' does not have a constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AnoConstructor)

File: src/Editor/Resources.hpp (hunk 2 above)
Comment (NinjaDanz3r) [inline src/Editor/Resources.hpp]: MINOR Unused private function: 'ResourceList::GetInstance' [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedPrivateFunction)

File: src/Editor/GUI/ResourceSelector.hpp (hunk 3 above)
Comment (NinjaDanz3r) [inline src/Editor/GUI/ResourceSelector.hpp]: MINOR The class 'ResourceSelector' does not have a constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AnoConstructor)

File: src/Editor/Resources.cpp (hunk 4 above)
Comment (NinjaDanz3r) [inline src/Editor/Resources.cpp]: MINOR Variable 'activeScene' is assigned in constructor body. Consider performing initialization in initialization list. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuseInitializationList)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 8 issues
* ![MINOR][MINOR] 8 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [SceneEditor.cpp#L20](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/2cfb99f3fd568b5b85d48b733db016c164e2705e/src/Editor/GUI/Editors/SceneEditor.cpp#L20): Member variable 'SceneEditor::sceneSelector' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/2cfb99f3fd568b5b85d48b733db016c164e2705e/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L559](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/2cfb99f3fd568b5b85d48b733db016c164e2705e/src/Engine/Manager/ScriptManager.cpp#L559): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/dee8e0468243170823e69a503492041fe543ba40/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L559](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/dee8e0468243170823e69a503492041fe543ba40/src/Engine/Manager/ScriptManager.cpp#L559): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d4614f3c586dbb42216aeca03cb51b09cc1e638d/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L559](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d4614f3c586dbb42216aeca03cb51b09cc1e638d/src/Engine/Manager/ScriptManager.cpp#L559): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**